### DISCUSSION TITLE
How to batch resize pictures keeping names and EXIF data

### DISCUSSION THREAD

**Discussion by tukoz** [ImageMagick/ImageMagick#7699] [Help] [answered]
### ImageMagick version

7.1.1-39

### Operating system, version and so on

Arch linux

### Description

Am looking for a reliable way to **resize** sets of JPG camera pictures to 1200×1200 pixels max, *preserving their filenames and EXIF data*. Some tips would be gladly appreciated, as for the hell of me I'm unable to do it.

Bash command used up to IM v6:
`for i in *.JPG; do convert -resize 1200x1200\> $i `basename $i .JPG`_small.jpg ; done`

Closest command I've been using since IM v7 -- but it removes EXIF data:
```ls *.JPG | xargs -I {} magick convert -quality 85 -thumbnail 1200 {} thumb.{}```

Below are the output obtained referring mostly to [Command-line Processing](https://imagemagick.org/script/command-line-processing.php) (Inline Image Resize, Image Geometry » Bounding the width, height, and area; the operators >, <, and @) as well I'm affraid to my own dumbiness: 

    $ magick -size 640x480 IMAGE_01.JPG IMAGE_01.jpg  # works

    $ magick IMAGE_02.JPG -resize '1200x1200>' IMAGE_02.jpg
    magick: insufficient image data in file `IMAGE_02.JPG' @ error/jpeg.c/ReadOneJPEGImage/1140

    $ magick IMAGE_02.JPG: -resize '1200x1200\>' IMAGE_02.jpg
    magick: unable to open image 'IMAGE_02.JPG:': No such file or directory @ error/blob.c/OpenBlob/3596.
    magick: no decode delegate for this image format `' @ error/constitute.c/ReadImage/746.

    $ magick '*.JPG[1200x1200>]' med%03d.jpg  # keep EXIF tags, but replace all images files names 

    $ magick image_01.JPG -resize 1200x1200>                                 
    bash: syntax error near unexpected token `newline

    $ magick mogrify -path webscale/ -resize 1200x1200> *.JPG                
    bash: *.JPG: ambiguous redirect

**Comment by dlemstra** [comment]
Can you share an image where this happens?

**Reply by tukoz** [reply]
Sure enough @dlemstra : [Bike_workshop_17.JPG](https://herouelibre-cloud.artifaille.fr/s/xPWWriyNLND2r7g)

**Reply by dlemstra** [reply]
That is not an image? Please zip the file and then drag and drop the zip file to this discussion.

**Reply by tukoz** [reply]
Explanation: the image I uploaded on our collective nextcloud was moved or renamed; sorry about that.

[atelier_Herouelibre_n75_Le-Bois_20241006_02.JPG.gz](https://github.com/user-attachments/files/17466554/atelier_Herouelibre_n75_Le-Bois_20241006_02.JPG.gz)
Note that only the last two attempts above, i.e. `$ magick image.JPG -resize 1200x1200> ` and `$ magick mogrify -path webscale/ -resize 1200x1200> *.JPG` do not work (with the same output) despite the image being the very same one.

Also, the main question has been considered for what it is: « How to *batch* resize pictures *keeping names and EXIF data* using IM v7+», yes? E.g.:
```bash
$ for i in *.JPG; do magick -resize '1200x1200\>' $i basename $i .JPG_small.jpg ; done
magick: no images found for operation `-resize' at CLI arg 1 @ error/operation.c/CLIOption/5481.
magick: no images found for operation `-resize' at CLI arg 1 @ error/operation.c/CLIOption/5481.
```

**Reply by dlemstra** [reply]
The order of your commands is incorrect you need to specify the file name before calling `-resize`.

**Reply by tukoz** [reply]
Considering the syntax given in the Command-line documentation `$ magick INPUT_IMAGE.JPG -resize '1200x1200>' DEST_IMAGE.jpg`
shouldn't it look like:
`$ for i in *.JPG; do magick $i -resize '1200x1200\>' basename $i .JPG_small.jpg; done`
?

I hope there was a way to batch rename, keeping the files original basename in an array. Something in the vein of:
`$ magick '*.jpg[120x120>]' thumbnail%03d.jpg` but with 'basename' in it.

**Comment by fmw42** [comment, marked as answer]
You can use (Unix syntax)

`magick *.jpg   -resize '1200x1200\>' -set filename:fname '%t' +adjoin '%[filename:fname]_small.jpg`

**Reply by tukoz** [reply]
Thank you @fmw42 as I did not find such an example. Might still miss something? (case sensitive enabled):
```bash
$ magick *.JPG -resize '1200x1200\>' -set filename:fname '%t'  '%[filename:fname].jpg'
magick: insufficient image data in file `atelier_Herouelibre_n77_Montmorency_20241020_04.JPG' @ error/jpeg.c/ReadOneJPEGImage/1140.
```
EDIT: removed the '+adjoin'

**Reply by fmw42** [reply]
The code is unix syntax. Change it appropriately for Windows.  Note, I removed the +adjoin and changed single quotes to double quotes.  Windows may want your % to double to %%.  Be sure you have enough RAM to hold all the images collected by the *.JPG

**Reply by tukoz** [reply]
This morning I took another two images of our last workshop and the command line you kindly provided works perfectly (just close the last quote). It also does exactly what we need, eg preserve the EXIF metadata :+1:  Now I can't remember encountering similar error outputs in almost 20 years using ImageMagick. Including over the past year using the same cameras.

> Operating system, version and so on
> 
> Arch linux

Note I can read (even if only out of respect for people taking the time to help) and even recognize Unix or Bash syntax ;-)  
There are two images in my testing directory, and 5-7 GiB available RAM (I also have 'zswap-cli' installed).

I'll do a bunch of tests to try to better pin point where the 'insufficient image data' problem comes from.

EDIT: I batch resized 600 pictures (including those used above) using @fmw42 's command with no error or warning *at all* .  
Maybe some command I used before trying the appropriate one yesterday (and before) caused the errors copy/pasted above.
